qutebrowser/app.py

```diff
@@ -837,6 +839,45 @@ def __init__(self, args):
         self.focusObjectChanged.connect(self.on_focus_object_changed)
         self.setAttribute(Qt.AA_UseHighDpiPixmaps, True)
 
+        self.new_window.connect(self._on_new_window)
+        config.instance.changed.connect(self._on_config_changed)
+
+    def _on_config_changed(self, option):
+        if option == 'tabs.undo_stack_size':
+            self._update_undo_stack_size()
+
+    def _on_new_window(self, window):
+        window.tabbed_browser.shutting_down.connect(
+            functools.partial(self._on_window_closing, window)
+        )
+
+    def _on_window_closing(self, window):
+        self._undos.append(mainwindow.WindowUndoEntry(
+            geometry=window.saveGeometry(),
+            private=window.tabbed_browser.is_private,
+            tab_stack=window.tabbed_browser.save_undo_stack(),
+        ))
+
+    def _update_undo_stack_size(self):
+        newsize = config.instance.get('tabs.undo_stack_size')
```

Comment:
Not a full review, but the docstring of `tabs.undo_stack_size` should probably be updated as well.

This also (as far as I can tell) increases the size of the largest undo stack much larger (100 deleted tabs stored in 100 deleted windows), but I don't think that case is very likely, as people are either going to use few windows and lots of tabs or the opposite.